### PR TITLE
Csproj format update, fixes #3944

### LINKED ISSUE #3944: Evaluate (and perform) csproj format update
The csproj files are a regular source of pain for pull-requests dealing with moving stuff around and just general development.  
MS did add a new csproj format that can use globbing and other cool (read: basic) features that are useful for remotely human readable project control.

The following guide can help migrate an old configuration to a new one: https://www.natemcmaster.com/blog/2017/03/09/vs2015-to-vs2017-upgrade/

### PR DESCRIPTION
**Pull request by Vogel612**

### CODE REVIEW

**Comment by bclothier** [comment]
Just few questions....

1) Was it intentional to delete [this inspection](https://github.com/rubberduck-vba/Rubberduck/pull/4383/files#diff-813398b2cd7dbff2cb0d3307ae86463e)
2) EasyHook files were apparently deleted. Is that intentional as well? I did not see a corresponding addition; only a `PackageReference`, and I am assuming we are now referencing the files in the `packages` folder which isn't a part of the git solution; but a bit unsure why files existed inside the projects originally. 
3) There were a number of conditions that were deleted that I don't think should be:
```
<Error Condition="!Exists('..\packages\Antlr4.CodeGenerator.4.6.4\build\Antlr4.CodeGenerator.props')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Antlr4.CodeGenerator.4.6.4\build\Antlr4.CodeGenerator.props'))" />
    <Error Condition="!Exists('..\packages\Antlr4.CodeGenerator.4.6.4\build\Antlr4.CodeGenerator.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Antlr4.CodeGenerator.4.6.4\build\Antlr4.CodeGenerator.targets'))" />
```
Judging from other diff, the `<Error Condition="..." />` tag seems legal even in the new csproj format. There were a number of projects that had those error condition for both ANTLR and NUnit.

**Comment by Vogel612** [comment]
Re 1: That inspection had been removed from the csproj files, but not from the git working tree. It uses old API for inspections and refers to resources that have been removed. The move to the new csproj format (which uses globbing to include files) just surfaced that mishap.

Re 2: EasyHook (as well as NLog.Schema and possibly other NuGet Packages) pushed parts of the content they should add to the `/packages` directory to the solution content. That behaviour is apparently not supported for `PackageReference` items. The changes in these cases were performed by the automatic NuGet package migration tool in VS. The tool produced a few warnings, but I haven't been able to spot any impact on the functionality in the final deployed assembly.

Re 3: Good spot, I'll make sure to reinstate these error conditions

**Comment by bclothier** [comment]
I believe I have discovered the source of instability in the build process. I think this section in the `Rubberduck.Core.csproj` is wrong.

```
  <!--BEGIN XAML WORKAROUND SECTION -->
  <ItemGroup>
    <Compile Include="$(IntermediateOutputPath)**\*.g.cs" Visible="false" />
    <Page Include="**\*.xaml">
      <SubType>Designer</SubType>
      <Generator>MSBuild:Compile</Generator>
      <!-- <Generator>MSBuild:UpdateDesignTimeXaml</Generator>-->
    </Page>
    <Compile Update="**\*.xaml.cs">
      <SubType>Code</SubType>
      <DependentUpon>%(Filename)</DependentUpon>
    </Compile>
    <!-- Resources -->
    <EmbeddedResource Update="Properties\Resources.resx" Generator="ResXFileCodeGenerator" LastGenOutput="Resources.Designer.cs" />
    <Compile Update="Properties\Resources.Designer.cs" AutoGen="True" DependentUpon="Resources.resx" DesignTime="True" />
    <!-- Settings -->
    <None Update="Properties\Settings.settings" Generator="SettingsSingleFileGenerator" LastGenOutput="Settings.Designer.cs" />
    <Compile Update="Properties\Settings.Designer.cs" AutoGen="True" DependentUpon="Settings.settings" />
  </ItemGroup>
  <!-- END XAML WORKAROUND SECTION -->
```

With this setting, the `obj` folder in the `Rubberduck.Core` goes nuts. Even when Visual Studio is just sitting there doing nothing, you can see files being modified, added, deleted nonstop in the `obj` folder. I think this triggers a infinite recursion trying to resolve all types in the intermediate files.

I made the following changes:
```
    <!--<Compile Include="$(IntermediateOutputPath)**\*.g.cs" Visible="false" />-->
    <Page Include="**\*.xaml">
      <SubType>Designer</SubType>
      <Generator>MSBuild:Compile</Generator>
      <!-- <Generator>MSBuild:UpdateDesignTimeXaml</Generator>-->
    </Page>
```

With this change, the `obj` folder became stable, Visual Studio starts performing better and build is now more consistent. I have not tested whether I have broke the WPF, however. I inferred the settings from [a SO thread](https://stackoverflow.com/questions/44140673/wpf-app-using-new-csproj-format).

With the build troubles out of the way, I was able to properly test and verify that yes, we _must_ have `[assembly: ComVisible(false)]`; the default is to make it COM-visible, which causes the MIDL to pick up unwanted types and thus fail with an error about missing forward declarations for certain types.

Please confirm if you'd like my commit to be added or if there's some other preference to merge the changes.

File: RubberduckBaseProject.csproj [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project>
```
Comment (bclothier) [inline RubberduckBaseProject.csproj]: This is easily my favorite feature of the new csproj format -- ability to standardize all projects. Love that! 👍

File: RubberduckBaseProject.csproj [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    <SolutionDir>$(MSBuildProjectDirectory)</SolutionDir>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <OutputPath>bin\$(Configuration)\</OutputPath>
+    <RunCodeAnalysis>true</RunCodeAnalysis>
+    <RestorePackages>true</RestorePackages>
+  </PropertyGroup>
+
+  <PropertyGroup Condition=" '$(AssemblyVersion)' == ''">
+    <AssemblyVersion>2.2.*</AssemblyVersion>
```
Comment (bclothier) [inline RubberduckBaseProject.csproj]: FWIW, this causes VS to emit a warning for each project about it being not a typical format.

File: RubberduckBaseProject.csproj [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    <FileVersion Condition=" '$(FileVersion)' == '' ">2.2.*</FileVersion>
+    <Deterministic>False</Deterministic>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Analyzer Condition=" '$(AssemblyName)' != 'RubberduckCodeAnalysis' "
```
Comment (bclothier) [inline RubberduckBaseProject.csproj]: The condition need to include `RubberduckTests` and `RubberduckCodeAnalysisTests`; makes no sense to run the analyzers there. (note: fixed in my commit that's not pushed yet)

File: RubberduckTestsCodeAnalysis/RubberduckTestsCodeAnalysis.csproj [hunk 4]
```diff
@@ -1,242 +1,160 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (bclothier) [inline RubberduckTestsCodeAnalysis/RubberduckTestsCodeAnalysis.csproj]: FYI for some reason the project isn't listed in the solution's object explorer.

File: Rubberduck.Resources/Rubberduck.Resources.csproj [hunk 5]
```diff
@@ -1,891 +1,29 @@
...
-      <DesignTime>True</DesignTime>
+  <Import Project="..\RubberduckBaseProject.csproj" />
+
+  <ItemGroup>
+    <Resource Include="**\*.png" />
+    <Resource Include="**\*.bmp" />
+    <Resource Include="**\*.txt" />
+    <Resource Include="**\*.resx" />
+    <Resource Update="**\*.de.resx">
+      <DependentUpon>$([System.String]::Copy('%(Filename)').Replace('.de', '')).resx</DependentUpon>
```
[the remark this answers is not in the document]
Comment (Vogel612) [reply Rubberduck.Resources/Rubberduck.Resources.csproj]: I'm not 100% clear on how wildcards work, but it might be worth a try to unify this, yea

**Comment by Vogel612** [comment]
Reminder to self: check whether we can use some of the MSBuild community tasks to simplify our stuff even more: https://github.com/loresoft/msbuildtasks

File: RubberduckTestsCodeAnalysis/RubberduckTestsCodeAnalysis.csproj (hunk 4 above)
Comment (Vogel612) [reply RubberduckTestsCodeAnalysis/RubberduckTestsCodeAnalysis.csproj]: This is fixed  with 74f09ba

**Comment by Vogel612** [comment]
Reminder to self: Update Analyzers and TestProject with https://stackoverflow.com/q/43602782/1803692 to simplify paths

File: RubberduckBaseProject.csproj (hunk 2 above)
Comment (Vogel612) [reply RubberduckBaseProject.csproj]: Added a comment about that. There's also something to note about how appveyor changes the version of the assemblies. It patches the AssemblyInfo.cs file of Rubberduck.Main. 
Unfortunately that no longer is a suitable option. I'll need to peruse AV docs on how to fix that.

File: RubberduckBaseProject.csproj [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    <OutputPath>bin\</OutputPath>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
+  </PropertyGroup>
+
+  <!-- Analyzer projects must set UnifyOutputPath to have this reference correctly -->
+  <Choose>
+    <When Condition=" '$(AssemblyName)' == 'RubberduckTests' ">
+      <ItemGroup>
+        <Analyzer Include="$(SolutionDir)\RubberduckTestsCodeAnalysis\bin\RubberduckTestsCodeAnalysis.dll"/>
```
Comment (bclothier) [inline RubberduckBaseProject.csproj]: That does not make sense - it seems to be saying that if the project is `RubberduckTests`, to add `RubberduckTestsCodeAnalysis` as an analyzer for that project. But the `RubberduckTestsCodeAnalysis` project is a test project testing the analzyer, not an analyzer project... Or am I misunderstanding?

**Comment by bclothier** [comment]
Regarding the problem with the WPF URI -- [this says](https://docs.microsoft.com/en-us/dotnet/framework/wpf/app-development/pack-uris-in-wpf) the version is optional. Do we really need to have the version embedded in the URI? Seems to me that it'd simplify lot of things if we only cared about the resource name, regardless of the version?

**Comment by Vogel612** [comment]
Reminder to self: get MSBump to put in versions: https://github.com/BalassaMarton/MSBump

**Comment by Vogel612** [comment]
Okay I think I'm done now. For me locally running the BuildRegistry script still fails from within VisualStudio. I haven't attempted to run it in msbuild alone. If I run the exact command that VS should run in a powershell host it works as expected and correctly registers the build.

As such I'm assuming that VS is somehow at fault. 

The problem with XAML has been symptomatically fixed, pending a more in-depth diagnosis of the underlying problem. I unfortunately am not remotely knowledgeable enough to find out where this goes wrong.

I have not included MSBump (yet?). Currently we're patching versions through AppVeyor, which works fine for AV. Locally the wildcard just goes to a generated buildnumber which is dependent on the timestamp. That way builds will properly generate different versions.

I have **not** verified that Fakes work as expected. Parsing, UnitTesting and similar things seem to work just as intended from cursory F5 testing. Unless there's something major, this is ready to merge from my side now.

File: RubberduckBaseProject.csproj (hunk 2 above)
Comment (retailcoder) [reply RubberduckBaseProject.csproj]: Is this resolved? How do we make RD v2.3.0?

**Pull request merged**